Step3-3

70ru5 opened the pull request:
## What
STEP3-3まで、とりあえず動くようになりました

（2/22 8:20追記）
みなさまからいただいたコメントをもとに、以下の修正を行いました。

・items.jsonのパスを定数として宣言しました。
・if文の書き方をショートステートメント付きにしました。
・エラーの返し方を具体的にしました。

（2/22 13:40追記）
・.DS_Storeを消して、下の記事を参考に.config/git/ignoreに.DS_Storeを書き込みました。

（2/23 0:26追記）
・saveImage関数を実装し、3-4の途中まで進みました。

（2/23 13:50追記）
・とりあえず最後まで実装しました。

## CHECKS :warning:

Please make sure you are aware of the following.

- [ ] **The changes in this PR doesn't have private information

70ru5 committed:
build environment

70ru5 committed:
step 3-2

70ru5 committed:
step 3-3

70ru5 committed:
step3-3

momom-i commented on go/items.json:
```diff
@@ -0,0 +1 @@
+[{"name":"jacket","category":"fashion"},{"name":"","category":""},{"name":"jacket","category":"fashion"},{"name":"","category":""},{"name":"jacket","category":"fashion"},{"name":"jacket","category":"fashion"}]
```
https://github.com/mercari-build/mercari-build-training/blob/main/document/03-api.ja.md#2-%E6%96%B0%E3%81%97%E3%81%84%E5%95%86%E5%93%81%E3%82%92%E7%99%BB%E9%8C%B2%E3%81%99%E3%82%8B

`{"items": [{"name": "jacket", "category": "fashion"}, ...]}`このようにitemsキーの配下に配列を記述するようにしてみましょうか！

70ru5 replied:
ありがとうございます。できたと思います！

参考にした記事のメモです
https://qiita.com/shuntaro_tamura/items/5a92c517a7d95d655505
https://qiita.com/chihiro/items/dd1912c1406dbfe16b72

momom-i commented on go/app/main.go:
```diff
@@ -20,22 +21,78 @@ type Response struct {
+		return err
+	}
+
+	err = os.WriteFile("items.json", newData, 0644)
+	if err != nil {
+		return err
+	}
```
```suggestion
        if err := os.WriteFile("items.json", newData, 0644); err != nil {
		return err
	}
```
このようにかけると綺麗です

task4233 replied:
imo: 他の場所ではerrのassignmentと比較部分を分けているので、一貫性を持たせるなら今のままでも良いと思います 👍 

(これは強い意見ではなく個人の意見なので、参考程度に解釈していただけると嬉しいです😃)

70ru5 replied:
ありがとうございます！ 修正しました。

momom-i commented on go/app/main.go:
```diff
@@ -20,22 +21,78 @@ type Response struct {
+	var items []Item
+	err = json.Unmarshal(data, &items)
+	if err != nil {
+		return err
```
93行目の正常系のレスポンスを返すときのように、エラーも同じ形で返してみましょう。[HTTPのステータスコード](https://developer.mozilla.org/ja/docs/Web/HTTP/Status)というのが定められていて、このステータスをサーバから返してもらうことで、クライアント側(curlとか叩いたりするときのリクエストを送る側の人)がどういったエラーなのか理解しやすくなります！
```go
c.JSON(http.StatusInternalServerError, err)
```

70ru5 replied:
ありがとうございます！ 修正しました。

質問や、確認したいことがあるのですが、お手隙の際にご対応いただければ幸いです。
・ファイルの読み込みやマーシャリングの際のエラーはサーバで行っている処理の失敗なのでステータスコードは500が適当だと思ったのですが、それで合っていますでしょうか？
・せっかくエラーメッセージを書いたので、エラーが起こったときどのように動くか試してみたいのですが、意図的にエラーを起こす方法は何かありますか？


以下は参考にした記事のメモです。

https://qiita.com/immrshc/items/13199f420ebaf0f0c37c
https://zenn.dev/syo_yamamoto/articles/2d9afd43617852
https://qiita.com/hayabusa_3288/items/0b8fab4a508d13543e0b

task4233 commented on .DS_Store:
nits: このファイルは不要なので、削除して.gitignoreに設定してあげると良いです 👍 
関係ないファイルを減らすとレビュアーの負荷が下がるので、他のメンバーのためになります

ちなみに、このファイルはmacOSのfinderなどでディレクトリを開くと自動で生成されるファイルです
開発とは関係ありません 🙆‍♂️

70ru5 replied:
知りませんでした！ ありがとうございます。

この辺り↓を見ながら設定しました！
https://qiita.com/inabe49/items/16ee3d9d1ce68daa9fff
https://zenn.dev/rabee/articles/gitignore-global-ds-store

task4233 commented on go/.DS_Store:
nits:同上

task4233 commented on go/app/main.go:
```diff
@@ -20,22 +21,78 @@ type Response struct {
+
 func root(c echo.Context) error {
-	res := Response{Message: "Hello, world!"}
+	res := Response{Message: "Hello, world!!"}
```
nits: 多分このPRと関係ないので戻しても良いかもです

70ru5 replied:
たぶんこれは変更を反映させる方法が分からなかったとき、変更が反映されてないのか単に動かないものを書いたのか確かめようと思って書いたものだと思います！

task4233 reviewed (commented):
実装お疲れ様です！
saveItemのように、一連の処理を関数にまとめてあげるのは可読性を高めるので良いと思います 👍 

ほか、気になったところだけコメントしましたが、あくまで些細なものなのでスルーしていただいても大丈夫です 🙆‍♂️

uta1018 commented on go/app/main.go:
```diff
@@ -20,22 +21,78 @@ type Response struct {
+func saveItem(name, category string) error {
+	item := Item{Name: name, Category: category}
+
+	data, err := os.ReadFile("items.json")
```
items.jsonのパスを定数として宣言すると、変更に強くなるのかなと思いました！

70ru5 replied:
ありがとうございます！ 修正しました。

uta1018 commented on go/app/main.go:
```diff
@@ -20,22 +21,78 @@ type Response struct {
+		return err
+	}
+
+	var items []Item
```
`var items []Item` が複数回宣言されているので、Itemと同様に構造体を定義してもよいのかなと思いました！

70ru5 replied:
ありがとうございます。修正しました！

uta1018 reviewed (commented):
2点コメントしました！
自分の理解なので、間違っていたらツッコミお願いします…！

cocodrips committed:
Add build.yml (#166)

## What

- Add build.yaml to workflow
- Resolve
https://github.com/mercari-build/mercari-build-training/issues/165


## CHECKS :warning:

Please make sure you are aware of the following.

- [x] **The changes in this PR doesn't have private information

70ru5 committed:
Add error messages

miaosakurai commented:
[nits] `.DS_Store`はリポジトリに入れる必要がないので、momom-sanがSlackに共有したようにgitignoreを使って`.DS_Store`をignoreすると便利です！
https://docs.github.com/ja/get-started/getting-started-with-git/ignoring-files
https://mercari.enterprise.slack.com/archives/C06JPTQPSQG/p1708496851935019

70ru5 committed:
Remove .DS_Store

70ru5 committed:
Update JSON file and made readItems function

miaosakurai commented on go/app/main.go:
```diff
@@ -83,20 +81,30 @@ func saveItem(name, category string) error {
-	return c.JSON(http.StatusOK, items)
+	if err = json.Unmarshal(data, &items); err != nil {
+		err := fmt.Errorf("error while unmarshaling file: %w", err)
+		return Items{}, err
```
```suggestion
return nil, err
```
nilでいい！

70ru5 committed:
modify error message

70ru5 committed:
add saveImage function

70ru5 committed:
add file writing function to func saveImage

70ru5 committed:
add filename to items

mayukoooo commented on go/app/main.go:
```diff
@@ -83,20 +81,30 @@ func saveItem(name, category string) error {
+	return c.JSON(http.StatusOK, items)
+}
+
+func readItems(filepath string) (Items, error) {
+	var items Items
+
+	data, err := os.ReadFile(filepath)
```
[nits]
非常に細かいですがキャメルですかね！👀
```suggestion
func readItems(filePath string) (Items, error) {
	var items Items

	data, err := os.ReadFile(filePath)
```

70ru5 replied:
ありがとうございます！ 修正しました！

mayukoooo commented on go/app/items.json:
@70ru5 
[nits][imo]
これはコミットする必要がないファイルなので削除で良いと思いますが、基本的にはファイルの末尾に改行が必要かな〜と思ったので一応コメントします🙏

[fyi]
- [なぜファイルの末尾に改行を入れたほうが良いのか - Qiita](https://qiita.com/hamamamama/items/11c17c7412a5aeb2ad74)
- [Why should text files end with a newline?](https://gist.github.com/OleksiyRudenko/d51388345ea55767b7672307fe35adf3)

70ru5 replied:
勉強になります。さっそく設定しました。ありがとうございます！

70ru5 committed:
3-5

70ru5 committed:
revise 3-5

dongurikoko commented on go/app/main.go:
```diff
@@ -13,29 +18,144 @@ import (
+		return "", err
+	}
+
+	fileName := fmt.Sprintf("%x", hash) + ".jpg"
```
[nits] 些細なことなのですが、Goでの "+" operatorを用いた文字列連結は低速みたいなので、fmt.Sprintf("%x.jpg", hash)みたいにまとめちゃうのが良さそうです！

70ru5 replied:
知りませんでした！ ありがとうございます。修正しました！
db3c7545d3ce8b760d01563c520be85d2b11e20d

dongurikoko commented on go/app/main.go:
```diff
@@ -13,29 +18,144 @@ import (
+	}
+
+	fileName := fmt.Sprintf("%x", hash) + ".jpg"
+	imagePath := "./images/" + fileName
```
同上で、fmt.Sprintfを使ってあげるといいかなと思います！

70ru5 replied:
こちらも修正しました！
db3c7545d3ce8b760d01563c520be85d2b11e20d

uta1018 reviewed (commented):
感想：
丁寧に役割ごとに関数が分けられていて、とてもよいなと思いました！
私も参考にさせていただきます…！

70ru5 committed:
fix variable name and string concatenation”
”

70ru5 committed:
change file permission

70ru5 committed:
add explanations of functions

task4233 commented on memo_STEP3:
imo: こういうまとめファイル、とても良いと思います！
syntax highlightがあった方が嬉しいと思うので、 `.txt` や `.md` のような拡張子を設定してあげると、より良いと思います 👍

task4233 commented on go/app/main.go:
```diff
@@ -51,6 +177,29 @@ func getImg(c echo.Context) error {
+func getInfo(c echo.Context) error {
+	itemId, err := strconv.Atoi(c.Param("id"))
+	if err != nil {
+		err := fmt.Errorf("invalid ID: %w", err)
+		return err
```
imo: ここで、Status codeを設定してあげるとクライアント側からエラー種別を判断できるので、より良いと思います :+1:
Shiori-sanの実装が綺麗だったので、良ければ参照してみてください 🙌 
https://github.com/dongurikoko/mercari-build-training/blob/main/go/app/main.go#L208-L211

task4233 reviewed (commented):
基本的にはLGTMです〜
いくつかコメントしたのですが、Status Codeの部分は対応した方が良いと思うので、対応をお願いします :pray:

それと、余裕があれが[3-6](https://github.com/70ru5/mercari-build-training/pull/2/files)もチャレンジしてみてください！

70ru5 committed:
modify error error handling, work on STEP 3-6, and add text files.
